server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java

```diff
@@ -411,83 +416,130 @@ public void initializeSnapshot(SnapshotId snapshotId, List<IndexId> indices, Met
+            }
             // Delete snapshot from the index file, since it is the maintainer of truth of active snapshots
-            final RepositoryData updatedRepositoryData = repositoryData.removeSnapshot(snapshotId);
-            writeIndexGen(updatedRepositoryData, repositoryStateId);
+            final RepositoryData repositoryData;
+            final RepositoryData updatedRepositoryData;
+            try {
+                repositoryData = getRepositoryData();
+                updatedRepositoryData = repositoryData.removeSnapshot(snapshotId);
+                writeIndexGen(updatedRepositoryData, repositoryStateId);
+            } catch (Exception ex) {
+                listener.onFailure(new RepositoryException(metadata.name(), "failed to delete snapshot [" + snapshotId + "]", ex));
+                return;
+            }
+            deleteSnapshotBlobs(snapshot, snapshotId, repositoryData, updatedRepositoryData, listener);
+        }
+    }
 
+    private void deleteSnapshotBlobs(SnapshotInfo snapshot, SnapshotId snapshotId, RepositoryData repositoryData,
+                                     RepositoryData updatedRepositoryData, ActionListener<Void> listener) {
+        final Executor executor = threadPool.executor(ThreadPool.Names.SNAPSHOT);
+        final AtomicInteger outstanding = new AtomicInteger(2);
+        final ActionListener<Void> deleteListener = ActionListener.wrap(() -> {
+            if (outstanding.decrementAndGet() == 0) {
+                deleteIndices(snapshot, repositoryData, snapshotId, updatedRepositoryData, listener);
+            }
+        });
+        executor.execute(() -> {
             // delete the snapshot file
             deleteSnapshotBlobIgnoringErrors(snapshot, snapshotId.getUUID());
+            deleteListener.onResponse(null);
+        });
+        executor.execute(() -> {
             // delete the global metadata file
             deleteGlobalMetaDataBlobIgnoringErrors(snapshot, snapshotId.getUUID());
+            deleteListener.onResponse(null);
+        });
+    }
 
-            // Now delete all indices
-            if (snapshot != null) {
-                final List<String> indices = snapshot.indices();
-                for (String index : indices) {
+    private void deleteIndices(SnapshotInfo snapshot, RepositoryData repositoryData, SnapshotId snapshotId,
+                               RepositoryData updatedRepositoryData, ActionListener<Void> listener) {
+        // Now delete all indices
+        if (snapshot != null) {
+            final List<String> indices = snapshot.indices();
+            if (indices.isEmpty()) {
+                deleteUnreferencedIndices(repositoryData, updatedRepositoryData, listener);
+                return;
+            }
+            final AtomicInteger outstandingIndices = new AtomicInteger(indices.size());
+            for (String index : indices) {
+                threadPool.executor(ThreadPool.Names.SNAPSHOT).execute(() -> {
```

Comment:
ActionRunnable

Reply:
We can further parallelize this to the shard level I think (can be a follow-up)

Reply:
@ywelsch yea I wanted to wait on #40250 for that, because we may want to push the specifics of that down to the blob container and handle the parallelization dependent on whether or not bulk deletes are supported.